src/GafferScene/EvaluateLightLinks.cpp

```diff
@@ -100,6 +111,98 @@ void EvaluateLightLinks::affects( const Gaffer::Plug *input, AffectedPlugsContai
 	}
 }
 
+void EvaluateLightLinks::hash( const Gaffer::ValuePlug *output, const Gaffer::Context *context, IECore::MurmurHash &h ) const
+{
+	if( output == lightNamesPlug() )
+	{
+		const StringData *expressionData = context->get<StringData>( g_expressionContextEntryName, nullptr );
+
+		if( expressionData )
+		{
+			h.append( SetAlgo::setExpressionHash( expressionData->readable(), inPlug() ) );
+			h.append( inPlug()->setHash( g_lightsSetName ) );
+		}
+		else
+		{
+			ScenePlug::SetScope scope( context, g_lightsSetName );
+
+			IECore::MurmurHash lightsSetHash = inPlug()->setPlug()->hash();
+			scope.setSetName( g_defaultLightsSetName );
+			IECore::MurmurHash defaultLightsSetHash = inPlug()->setPlug()->hash();
+			if( lightsSetHash == defaultLightsSetHash )
+			{
+				h.append( 0 );
+				return;
+			}
+
+			ConstPathMatcherDataPtr defaultLightsSetData = inPlug()->setPlug()->getValue( &defaultLightsSetHash );
+			scope.setSetName( g_lightsSetName );
+			ConstPathMatcherDataPtr lightsSetData = inPlug()->setPlug()->getValue( &lightsSetHash );
+			if( lightsSetData->readable() == defaultLightsSetData->readable() )
+			{
+				h.append( 0 );
+				return;
+			}
+
+			h.append( lightsSetHash );
+			h.append( defaultLightsSetHash );
+		}
+
+		return;
+	}
+
+	SceneProcessor::hash( output, context, h );
+}
+
+void EvaluateLightLinks::compute( Gaffer::ValuePlug *output, const Gaffer::Context *context ) const
+{
+	if( output == lightNamesPlug() )
+	{
+		const StringData *expressionData = context->get<StringData>( g_expressionContextEntryName, nullptr );
+
+		ObjectPtr lightsObject = NullObject::defaultNullObject();
+		{
+			ScenePlug::SetScope scope( context, g_lightsSetName );
+
+			IECore::MurmurHash lightsSetHash = inPlug()->setPlug()->hash();
+			ConstPathMatcherDataPtr lightsSetData = inPlug()->setPlug()->getValue( &lightsSetHash );
+			const PathMatcher &lightsSet = lightsSetData->readable();
+
+			if( expressionData )
+			{
+				lightsObject = expressionToLightNames( expressionData->readable(), inPlug(), lightsSet );
+			}
+			else
+			{
+				scope.setSetName( g_defaultLightsSetName );
+				IECore::MurmurHash defaultLightsSetHash = inPlug()->setPlug()->hash();
+
+				if( defaultLightsSetHash != lightsSetHash )
+				{
+					ConstPathMatcherDataPtr defaultLightsSetData = inPlug()->setPlug()->getValue( &defaultLightsSetHash );
+					const PathMatcher &defaultLightsSet = defaultLightsSetData->readable();
+
+					if( lightsSet != defaultLightsSet )
+					{
+						PathMatcher defaultLightsIntersected = defaultLightsSet.intersection( lightsSet );
```

Comment:
So, I've changed that little helper, now called `setToValidLightNames`, to return an `ObjectPtr` based on two `PathMatchers`. That way it's actually helping with DRY. By only passing in the two sets (instead of a scene plug as before), it is pretty obvious which data it depends on, I think. Even though there's some indirection, it helps with keeping the structure of hash and compute pretty identical.